### ISSUE TITLE
DirectoryInfo.CreateSubdirectory fails in .NET Core 2.1 when DirectoryInfo path ends in separator

### ISSUE THREAD

**Issue by daxian-dbw** [dotnet/corefx#30283]
### Repro Steps:
```c#
        static void Main(string[] args)
        {
            var d = new DirectoryInfo(@"C:\");
            d.CreateSubdirectory("blah");
        }
```

### Fail when building with `netcoreapp2.1`
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>

</Project>
```

```none
## version 2.1.300
> dotnet.exe run

Unhandled Exception: System.ArgumentException: The directory specified, 'blah', is not a subdirectory of 'C:\'.
Parameter name: path
   at System.IO.DirectoryInfo.CreateSubdirectory(String path)
   at dumpConsole.Program.Main(String[] args) in E:\arena\dotnetApp\test\dumpConsole\Program.cs:line 11
```

### Work when building with `netcoreapp2.0`
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

</Project>
```
```none
## version 2.1.300
> dotnet.exe run

## the folder "c:\blah" was created successfully.
```

This regression might be introduced by #27810

**Comment by danmosemsft** [comment]
@JeremyKuhne this is a regression. The bug seems to be here

https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs#L67
```c#
            if (fullPath.Length < FullPath.Length 
                || (fullPath.Length > FullPath.Length && !PathInternal.IsDirectorySeparator(fullPath[FullPath.Length]))
                || string.Compare(FullPath, 0, fullPath, 0, FullPath.Length, PathInternal.StringComparison) != 0)
            {
                throw new ArgumentException(SR.Format(SR.Argument_InvalidSubPath, path, FullPath), nameof(path));
            }
```

This check `!PathInternal.IsDirectorySeparator(fullPath[FullPath.Length]))` will fail if the parent directory passed in ends with a slash, because FullPath.Length will point beyond the slash. The same happens if I try to make a child folder of `c:\foo\` (ending with slash)

**Comment by danmosemsft** [comment]
In 2.0 it just did
```C#
            if (0 != String.Compare(FullPath, 0, fullPath, 0, FullPath.Length, PathInternal.StringComparison))
            {
                throw new ArgumentException(SR.Format(SR.Argument_InvalidSubPath, path, DisplayPath), nameof(path));
            }
```

**Comment by danmosemsft** [comment]
Introduced in 59f6987b7fc919a438060f6c502c3c9cf7c8af29

**Comment by JeremyKuhne** [comment]
Ugh, we're clearly missing a test here. I'll fix it shortly. Thanks for filing the issue, @daxian-dbw!

**Comment by karelz** [comment]
This has been fixed in release/2.1 branch in PR #30367. Closing.
@JeremyKuhne can you please double check?
